tests/test_jinja_templates.py

```diff
@@ -0,0 +1,216 @@
+            + urllib.parse.quote("a_token")
+            + "%26private_key%3D"
+            + urllib.parse.quote("mystery_key")
+            + "&room_name="
+            + urllib.parse.quote("Bob's Emporium of Messages")
+            + "&room_avatar_url="
+            + urllib.parse.quote("mxc://example.org/s0meM3dia")
+            + "&inviter_name="
+            + urllib.parse.quote("<Bob Smith>")
+            + "&guest_access_token=&guest_user_id="
+        )
+        text = email_contents.splitlines()
+        link = text[22]
+        self.assertEqual(link, expected_url)
+
+    def test_jinja_matrix_verification(self):
+        substitutions = {
+            "address": "foo@example.com",
+            "medium": "email",
+            "to": "person@test.test",
+            "token": "<<token>>",
+            "link": "https://link_test.com",
+        }
+
+        templateFile = self.sydent.get_branded_template(
+            "matrix-org",
+            "verification_template.eml",
+            ("email", "email.verification_template"),
+        )
+
+        with patch("sydent.util.emailutils.smtplib") as smtplib:
+            sendEmail(self.sydent, templateFile, "test@test.com", substitutions)
+
+        smtp = smtplib.SMTP.return_value
+        email_contents = smtp.sendmail.call_args[0][2].decode("utf-8")
+
+        # test url input is encoded
+        self.assertIn(urllib.parse.quote("https://link_test.com"), email_contents)
+
+        # test html input is escaped
+        self.assertIn("&lt;&lt;token&gt;&gt;", email_contents)
+
+        # test safe values are not escaped
+        self.assertIn("<<token>>", email_contents)
+
+    def test_jinja_vector_verification(self):
+        substitutions = {
+            "address": "foo@example.com",
+            "medium": "email",
+            "to": "person@test.test",
+            "link": "https://link_test.com",
+        }
+
+        templateFile = self.sydent.get_branded_template(
+            "vector-im",
+            "verification_template.eml",
+            ("email", "email.verification_template"),
+        )
+
+        with patch("sydent.util.emailutils.smtplib") as smtplib:
+            sendEmail(self.sydent, templateFile, "test@test.com", substitutions)
+
+        smtp = smtplib.SMTP.return_value
+        email_contents = smtp.sendmail.call_args[0][2].decode("utf-8")
+
+        # test url input is encoded
+        self.assertIn(urllib.parse.quote("https://link_test.com"), email_contents)
+
+        path = os.path.join(
+            self.sydent.cfg.get("general", "templates.path"),
+            "vector_verification_sample.txt",
+        )
+
+        with open(path, "r") as file:
+            expected_text = file.read()
+
+        # remove headers as they are variable
+        email_contents = email_contents.splitlines()[13:]
+
+        # remove multipart headers as they are variable
+        del email_contents[40]
+        del email_contents[156]
```

Comment:
I understand why you've done that, but especially by referring to numeric indices, this could be a bit brittle (if we change the template a bit, the test will break until we tweak these numbers).
Is it perhaps possible to patch out the multipart boundary generating function with one that generates the same boundary each time?
An alternative would be to pre-seed the random number generator so that you know which things it will emit in which order (though that looks trickier than just patching `generateAlphanumericTokenOfLength` to be deterministic, because it seems to use the system's generator).